language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,615 @@
+            -- separately check the VarkInds environment for conflicts
+            assertNoRecConflict . mapMaybe bindingNewTyId $ toList bs
+            -- now with all bindings in the env, check for wellformedness
+            withBinds bs $ do
+              checkWellformBinds bs
+              inferTypeM inTerm
+    -- In paper's Fig.6, T-Let and T-LetRec rules dictate that: G !- inTerm :: *
+    -- Durining implemenetation, however, this leads to scoping errors because
+    -- the G env is not augmented to see the kinds introduced by any (nested) let(s)
+    -- So we disable this check: `checkKindM ann (ann <$ unNormalized tyInTerm) $ Type ()`
+    -- Furthermore, this check would be unnecessary afterall, because of the proof-by-construction:
+    -- All terms that can be constructed have types that are *-kinded.
+    pure tyInTerm
+
+  where
+    withBinds :: NonEmpty (Binding TyName Name uni ann) -> TypeCheckM uni ann a -> TypeCheckM uni ann a
+    withBinds = flip . foldr $ withBind
+
+    checkWellformBinds :: NonEmpty (Binding TyName Name uni ann) -> TypeCheckM uni ann ()
+    checkWellformBinds = traverse_ $ checkWellformBind Rec
+
+    assertNoRecConflict :: [T.Text] -> TypeCheckM uni ann ()
+    assertNoRecConflict = maybe (pure ())
+                          (throwError . ConflictingRecBinds ann1)
+                           . hasDuplicate
+
+    -- TODO: duplicate code, also move to PlutusIR.hs
+    bindingNewIds :: Binding TyName Name uni ann -> [T.Text]
+    bindingNewIds = \case
+        TermBind _ _ d _ -> [nameString $ varDeclName d]
+        DatatypeBind _ (Datatype _ _ _ des vdecls) ->
+            nameString des : fmap (nameString . varDeclName) vdecls
+        _typebind -> []
+
+    -- a binding may introduce max. 1 new id in scope, hence maybe
+    -- TODO: duplicate code, also move to PlutusIR.hs
+    bindingNewTyId :: Binding TyName Name uni ann -> Maybe T.Text
+    bindingNewTyId = \case
+        TypeBind _ tvdecl _ ->
+            pure . nameString . unTyName $ tyVarDeclName tvdecl
+        DatatypeBind _ (Datatype _ tvdecl _ _ _) ->
+            pure . nameString . unTyName $ tyVarDeclName tvdecl
+        _termbind -> mempty
+
+-- See the [Global uniqueness] and [Type rules] notes.
+-- | Check a 'Term' against a 'NormalizedType'.
+checkTypeM
+    :: (GShow uni, GEq uni, DefaultUni <: uni)
+    => ann -> Term TyName Name uni ann -> Normalized (Type TyName uni ()) -> TypeCheckM uni ann ()
+
+-- [infer| G !- term : vTermTy]    vTermTy ~ vTy
+-- ---------------------------------------------
+-- [check| G !- term : vTy]
+checkTypeM ann term vTy = do
+    vTermTy <- inferTypeM term
+    when (vTermTy /= vTy) $ throwError (TypeMismatch ann (void term) (unNormalized vTermTy) vTy)
+
+-- | Extends the environment(s) with a bind.
+-- NOTE: there are no well-formed or duplication checks here, it just adds to the environment(s) like withVar/withTyVar do.
+-- For checks see `checkWellformBind`.
+withBind :: forall uni ann a.
```

Comment:
I'd say `withBinding`. I'm not sure we abbreviate "binding" to "bind" generally.

Reply:
I think I am brainwashed by ghc (e.g. `-fno-warn-unused-binds`) :D